h/static/scripts/websocket.js

```diff
@@ -32,44 +38,80 @@ function Socket(url) {
     }
   }
 
-  function reconnect() {
-    var didConnect = false;
-    var connectOperation = retry.operation();
-    connectOperation.attempt(function (currentAttempt) {
+  // Connect the websocket immediately. If a connection attempt is already in
+  // progress, do nothing.
+  function connect() {
+    if (operation) {
+      return;
+    }
+
+    operation = retry.operation({
+      minTimeout: RECONNECT_MIN_DELAY * 2,
+      // Don't retry forever -- fail permanently after 10 retries
+      retries: 10,
+      // Randomize retry times to minimise the thundering herd effect
+      randomize: true
+    });
+
+    operation.attempt(function () {
       socket = new WebSocket(url);
       socket.onopen = function (event) {
-        // signal successful connection
-        connectOperation.retry();
-        didConnect = true;
-        sendMessages();
-
+        onOpen();
```

Comment:
I find it a little confusing here that some of the logic for the open event is in `onOpen()` and some is in the 'onopen' handler.